Add aet and spr db support

vixen256 opened the pull request.

vixen256 committed:
Add aet and spr db support (No python FFI)

Waelwindows committed:
refactor: Move readers for aet/spr to read mod

Move the readers into their own module in order to fit in with the rest
of the codebase

Waelwindows committed:
refactor(aet): Use BTreeMap instead of Vec

This should reflect how the db is actually used. (acces via id not by index).
As a compromise, `index` is stored to preserve order from the original
for special needs (personal communications with Brogamer)

Waelwindows committed:
refactor(spr): Use BTreeMap instead of Vec

This should reflect how the db is actually used. (acces via id not by index).
As a compromise, `index` is stored to preserve order from the original
for special needs (personal communications with Brogamer)

Waelwindows committed:
refactor(spr/aet): Use readers for parsing

The responsibilty of providing the bytes should be on the user.
This also lowers the API ground so that in the future `no_std` is possible

Waelwindows committed:
test(aet): Add test case for reading

Waelwindows committed:
refactor(aet): Encapsulate conversion for read

This should make the code easier to read and make it easier to convert
to and from the reader types for writing in the future

Waelwindows committed:
refactor(aet/spr): Rename `read` module to `io`

This is to anticipate future serialization support

Waelwindows committed:
test(spr): Add read test case

Waelwindows committed:
feat(spr): Implement common traits for structs

Waelwindows commented:
How do you think we should handle new db entries? Specifically with respect to `index`?
AFAICU the `index` field is used just maintain order in the db, no? 
If that's the case, perhaps we could map it as an `Option<u32>` so that new entries are automatically at the end of the db

Waelwindows committed:
refactor(spr/io): Encapsulate type conversion

This should help in the future when we support serialization

Waelwindows committed:
feat(spr): Add `serde` support

Waelwindows committed:
feat(spr/py_ffi): Add python ffi

Waelwindows committed:
feat(aet): Implement serde

Waelwindows committed:
feat(aet): Implement common traits

Waelwindows committed:
feat(aet/py_ffi): Add python ffi

Waelwindows committed:
feat(aet/py_ffi): Add `get_spr_set` helper

This should help python ffi users to get the corresponding spr set safely

Waelwindows committed:
feat(aet/spr): Expose `index`

Personal correspondings with Brogamer suggseted that `index` field should
be visible publicly. It also simplifies our API as we don't have to
maintain the invariant of index being private

Waelwindows committed:
feat(spr/aet): Implement `Default` for structs

This goes hand in hand with the previous commit, fully making the struct
plain old data structs

Waelwindows committed:
build(nix): Update lock

`binrw` did not compile with the used version inside of nixos builders.
Updating to the latest unstable build fixes this

Waelwindows committed:
build(nix): Add flake checks

Waelwindows committed:
feat(aet/spr): Add `from_bytes` ergonomic wrapper

Helps keep in line with the original function signatures of the PR.
Also less boilerplate for reading slices.

Waelwindows committed:
doc(spr): Add documentation

Waelwindows committed:
doc(aet): Add documentation

Waelwindows committed:
perf(aet/spr/py_ffi): Use `PyBytes` in `read`

Using `PyBytes` avoids the cost of converting the list in python

See
https://github.com/PyO3/pyo3/issues/100#issuecomment-1220672112

Waelwindows committed:
chore(py_ffi): Update `repr` to include quotes

This make it more clear what the name is even when it is empty

Waelwindows committed:
feat(aet/spr/py_ffi): Add constructors

Waelwindows committed:
doc(aet/spr/doc): Add documentation

Should help consumers from both sides understand how to use the APi

Waelwindows committed:
docs(py_ffi): Add module level documentation
